src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+#else
+#define request_t int
+#define comm_t int
+#endif
+
+namespace parthenon {
+
+//             Read    Write
+//    stale:             X
+// sending*:
+// received:     X
+enum class BufferState { stale, sending, sending_null, received, received_null };
+
+enum class BuffCommType { sender, receiver, both, sparse_receiver };
+
+enum class BoundaryType : int { local, nonlocal, any, reflux_send, reflux_recv };
+
+template <class T>
+class CommBuffer {
+ private:
+  // Need specializations to be friends with each other
+  template <typename U>
+  friend class CommBuffer;
+
+  std::shared_ptr<BufferState> state_;
+  std::shared_ptr<BuffCommType> comm_type_;
+  std::shared_ptr<bool> started_irecv_;
+  std::shared_ptr<int> nrecv_tries_;
+  std::shared_ptr<request_t> my_request_;
+
+  int my_rank;
+  int tag_;
+  int send_rank_;
+  int recv_rank_;
+  comm_t comm_;
+
+  using buf_base_t = std::remove_pointer_t<decltype(std::declval<T>().data())>;
+  buf_base_t null_buf_ = std::numeric_limits<buf_base_t>::signaling_NaN();
+  bool active_ = false;
+
+  std::function<T()> get_resource_;
+
+  T buf_;
+
+ public:
+  CommBuffer()
+      : my_rank(0)
+#ifdef MPI_PARALLEL
+        ,
+        my_request_(std::make_shared<MPI_Request>(MPI_REQUEST_NULL))
+#endif
+  {
+  }
+
+  CommBuffer(int tag, int send_rank, int recv_rank, comm_t comm_,
+             std::function<T()> get_resource, bool do_sparse_allocation = false);
+
+  ~CommBuffer();
+
+  template <class U>
+  CommBuffer(const CommBuffer<U> &in);
+
+  template <class U>
+  CommBuffer &operator=(const CommBuffer<U> &in);
+
+  operator T &() { return buf_; }
+  operator const T &() const { return buf_; }
+
+  T &buffer() { return buf_; }
+  const T &buffer() const { return buf_; }
+
+  void Allocate() {
+    if (!active_) {
+      buf_ = get_resource_();
+      active_ = true;
+    }
+  }
+
+  void Free() {
+    buf_ = T();
+    active_ = false;
+  }
+
+  bool IsActive() const { return active_; }
+
+  BufferState GetState() { return *state_; }
+
+  void Send() noexcept;
+  void SendNull() noexcept;
+
+  bool IsAvailableForWrite();
+
+  void TryStartReceive() noexcept;
+  bool TryReceive() noexcept;
+
+  void Stale();
+};
+
+// Method definitions below
+
+template <class T>
+CommBuffer<T>::CommBuffer(int tag, int send_rank, int recv_rank, comm_t comm,
+                          std::function<T()> get_resource, bool do_sparse_allocation)
+    : state_(std::make_shared<BufferState>(BufferState::stale)),
+      comm_type_(std::make_shared<BuffCommType>(BuffCommType::both)),
+      started_irecv_(std::make_shared<bool>(false)),
+      nrecv_tries_(std::make_shared<int>(0)),
+#ifdef MPI_PARALLEL
+      my_request_(std::make_shared<MPI_Request>(MPI_REQUEST_NULL)),
+#endif
+      tag_(tag), send_rank_(send_rank), recv_rank_(recv_rank), comm_(comm),
+      get_resource_(get_resource), buf_() {
+// Set up persistent communication
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
```

Comment:
Why not use `Globals::my_rank` here?
This would also remove the #ifdef` as it's set to 0 for serial runs.

Reply:
Yep, that is a good suggestion.

Reply:
done.